lighthouse-core/gather/gatherers/seo/font-size.js

```diff
@@ -0,0 +1,166 @@
+    .then(result => {
+      const body = findBody(result.root);
+      return body ? flattenTree(body) : [];
+    });
+}
+
+/**
+ * Returns effective CSS rule for given CSS property
+ *
+ * @param {!string} property CSS property name
+ * @param {!Node} node
+ * @param {!Object} matched CSS rules
+ * @returns {WebInspector.CSSStyleDeclaration}
+ */
+function getEffectiveRule(property, node, {
+  inlineStyle,
+  attributesStyle,
+  matchedCSSRules,
+  inherited,
+}) {
+  const cssModel = {
+    styleSheetHeaderForId: id => ({id}),
+  };
+
+  const nodeType = node.nodeType;
+  node.nodeType = () => nodeType;
+  const matchedStyles = new CSSMatchedStyles(
+    cssModel,
+    node,
+    inlineStyle,
+    attributesStyle,
+    matchedCSSRules,
+    null,
+    inherited,
+    null
+  );
+
+  const nodeStyles = matchedStyles.nodeStyles();
+  const matchingRule = nodeStyles.find(style => {
+    const foundProperty = style.allProperties.find(item => item.name === property);
+    return foundProperty &&
+      matchedStyles.propertyState(foundProperty) !== CSSMatchedStyles.PropertyState.Overloaded;
+  });
+
+  return matchingRule;
+}
+
+/**
+ * @param {!Object} driver
+ * @param {!Node} node text node
+ * @returns {!{fontSize: number, textLength: number, node: Node, cssRule: WebInspector.CSSStyleDeclaration}}
+ */
+function getFontSizeInformation(driver, node) {
+  const computedStyles = driver.sendCommand('CSS.getComputedStyleForNode', {nodeId: node.parentId});
+  const matchedRules = driver.sendCommand('CSS.getMatchedStylesForNode', {nodeId: node.parentId});
+
+  return Promise.all([computedStyles, matchedRules])
+    .then(result => {
+      const [{computedStyle}, matchedRules] = result;
+      const fontSizeProperty = computedStyle.find(({name}) => name === FONT_SIZE_PROPERTY_NAME);
+
+      return {
+        fontSize: parseInt(fontSizeProperty.value, 10),
+        textLength: node.nodeValue.trim().length,
+        node: node.parentNode,
+        cssRule: getEffectiveRule(FONT_SIZE_PROPERTY_NAME, node.parentNode, matchedRules),
+      };
+    });
+}
+
+/**
+ * @param {Node} node
+ * @returns {boolean}
+ */
+function isNonEmptyTextNode(node) {
+  return node.nodeType === global.Node.TEXT_NODE && node.nodeValue.trim().length > 0;
+}
+
+class FontSize extends Gatherer {
+  /**
+   * @param {{driver: !Object}} options Run options
+   * @return {!Promise<Array<{fontSize: number, textLength: number, node: Node, cssRule: WebInspector.CSSStyleDeclaration}>>} font-size analysis
+   */
+  afterPass(options) {
+    const enableDOM = options.driver.sendCommand('DOM.enable');
+    const enableCSS = options.driver.sendCommand('CSS.enable');
+
+    return Promise.all([enableDOM, enableCSS])
+      .then(() => getAllNodesFromBody(options.driver))
+      .then(nodes => nodes.filter(isNonEmptyTextNode))
+      .then(textNodes => Promise.all(
+        textNodes.map(node => getFontSizeInformation(options.driver, node))
```

Comment:
this seems like it could have 1000s of inflight commands, 2 for each node in the body right? 

any thoughts on limiting the protocol round-trips to try and compute as much as possible on browser side or do we absolutely need the node IDs for attribution :/

Reply:
> 2 for each node in the body 
For each non empty text node in the body, yes.

Yeah, I was also worried about that part, but interestingly it performs very well. I understand your concern though. If we would like to keep the number of sent commands down I do have some ideas:

- get all computed styles in one shot with `DOMSnapshot.getSnapshot` (we should be able to connect info from that method with info from `DOM.getDocument` via backendNodeId )
- inject a script that will collect font-size information in the browser context (but it will be tricky to connect that info with Node objects from `DOM.getDocument`)
- only call `getMatchedStylesForNode` for nodes that have font-size below the threshold (but this would mean making the gatherer less generic)
- filter nodes a bit more to only consider visible ones (figuring out what's visible might be tricky though)

Let me know what you think!

Reply:
> but interestingly it performs very well.

alright then works for me as is :) maybe throw a comment in there saying it works out fine because X so future people don't bother trying to fix?

Reply:
I did some testing (using [url](https://www.flynashville.com/Pages/default.aspx) that you provided):

## current solution
getting nodes: 631.190ms
getting font size info: 6445.445ms
whole thing: 7212.822ms

## w/o CSS.getComputedStyleForNode
getting nodes: 717.872ms
getting font size info: 6005.998ms
whole thing: 6726.438ms

## w/o CSS.getMatchedStylesForNode
getting nodes: 621.086ms
getting font size info: 215.401ms
whole thing: 839.306ms

## CSS.getMatchedStylesForNode only for nodes with font-size < 16px
getting nodes: 576.719ms
getting font size info: 4676.480ms
whole thing: 5280.742ms

It looks like `CSS.getMatchedStylesForNode` is solely responsible for bad performance and calling it only for nodes with font-size below the threshold doesn't change much for pages like https://www.flynashville.com/Pages/default.aspx .

Ideas that we had were around getting nodes and computed styles for them. However, this doesn't seem to be a bottleneck. We will need some ideas for dealing with `CSS.getMatchedStylesForNode`. Calling it only for nodes below the threshold doesn't do the trick.

Wish there was something like https://chromedevtools.github.io/devtools-protocol/tot/CSS/#method-setEffectivePropertyValueForNode but for getting effective rules for given property. Or ability to make more specific `CSS.getMatchedStylesForNode` calls (e.g. tell that we only care about one property and only about effective rule).

Reply:
As discussed with @patrickhulce right now - we will only call `getMatchedStylesForNode` for top X nodes with longest text. If website has more than X failing nodes user will get information at the end of the table that this is a partial result. I'll try to figure out what X value is to keep this audit under a second.